File: SMILE/tutorials/cpp/tutorial1.cpp
// tutorial1.cpp
// Tutorial1 creates a simple network with three nodes,
// then saves it as XDSL file to disk.

#include "smile.h"
#include <cstdio>


static int CreateCptNode(
    DSL_network &net, const char *id, 
    const char *name, const char *outcomes[], 
    int xPos, int yPos);


int Tutorial1()
{
    printf("Starting Tutorial1...\n");

    // show errors and warnings in the console
    DSL_errorH().RedirectToFile(stdout);

    DSL_network net;

    const char *ECONOMY_OUTCOMES[] = { "Up", "Flat", "Down", NULL };
    int e = CreateCptNode(net, "Economy", "State of the economy", 
        ECONOMY_OUTCOMES, 160, 40);

    const char *SUCCESS_OUTCOMES[] = { "Success", "Failure", NULL };
    int s = CreateCptNode(net, "Success", "Success of the venture", 
        SUCCESS_OUTCOMES, 60, 40);
    
    const char *FORECAST_OUTCOMES[] = { "Good", "Moderate", "Poor", NULL };
    int f = CreateCptNode(net, "Forecast", "Expert forecast", 
        FORECAST_OUTCOMES, 110, 140);

    net.AddArc(e, s);
    net.AddArc(s, f);
    net.AddArc(e, f);

    DSL_nodeDefinition *economyDef = net.GetNode(e)->Definition();
    DSL_doubleArray ep(economyDef->GetMatrix()->GetSize());
    ep[0] = 0.2; // P(Economy=U)
    ep[1] = 0.7; // P(Economy=F)
    ep[2] = 0.1; // P(Economy=D)
    int res = economyDef->SetDefinition(ep);
    if (DSL_OKAY != res)
    {
        return res;
    }

    DSL_nodeDefinition *successDef = net.GetNode(s)->Definition();
    DSL_doubleArray sp(successDef->GetMatrix()->GetSize());
    sp[0] = 0.3; // P(Success=S|Economy=U)
    sp[1] = 0.7; // P(Success=F|Economy=U)
    sp[2] = 0.2; // P(Success=S|Economy=F)
    sp[3] = 0.8; // P(Success=F|Economy=F)
    sp[4] = 0.1; // P(Success=S|Economy=D)
    sp[5] = 0.9; // P(Success=F|Economy=D)
    res = successDef->SetDefinition(sp);
    if (DSL_OKAY != res)
    {
        return res;
    }

    DSL_nodeDefinition *forecastDef = net.GetNode(f)->Definition();
    DSL_doubleArray fp(forecastDef->GetMatrix()->GetSize());
    fp[0] = 0.70; // P(Forecast=G|Success=S,Economy=U)
    fp[1] = 0.29; // P(Forecast=M|Success=S,Economy=U)
    fp[2] = 0.01; // P(Forecast=P|Success=S,Economy=U)
    
    fp[3] = 0.65; // P(Forecast=G|Success=S,Economy=F)
    fp[4] = 0.30; // P(Forecast=M|Success=S,Economy=F)
    fp[5] = 0.05; // P(Forecast=P|Success=S,Economy=F)

    fp[6] = 0.60; // P(Forecast=G|Success=S,Economy=D)
    fp[7] = 0.30; // P(Forecast=M|Success=S,Economy=D)
    fp[8] = 0.10; // P(Forecast=P|Success=S,Economy=D)

    fp[9] = 0.15;  // P(Forecast=G|Success=F,Economy=U)
    fp[10] = 0.30; // P(Forecast=M|Success=F,Economy=U)
    fp[11] = 0.55; // P(Forecast=P|Success=F,Economy=U)

    fp[12] = 0.10; // P(Forecast=G|Success=F,Economy=F)
    fp[13] = 0.30; // P(Forecast=M|Success=F,Economy=F)
    fp[14] = 0.60; // P(Forecast=P|Success=F,Economy=F)

    fp[15] = 0.05; // P(Forecast=G|Success=F,Economy=D)
    fp[16] = 0.25; // P(Forecast=G|Success=F,Economy=D)
    fp[17] = 0.70; // P(Forecast=G|Success=F,Economy=D)

    res = forecastDef->SetDefinition(fp);
    if (DSL_OKAY != res)
    {
        return res;
    }

    res = net.WriteFile("tutorial1.xdsl");
    if (DSL_OKAY != res)
    {
        return res;
    }

    printf("Tutorial1 complete: Network written to tutorial1.xdsl\n");
    return DSL_OKAY;
}


static int CreateCptNode(
	DSL_network &net, const char *id, const char *name, 
    const char *outcomes[], int xPos, int yPos)
{
    int handle = net.AddNode(DSL_CPT, id);
    DSL_node *node = net.GetNode(handle);

    DSL_idArray ida;
    for (const char **p = outcomes; *p != NULL; p++)
    {
        ida.Add(*p);
    }
    DSL_nodeDefinition *def = node->Definition();
    def->SetNumberOfOutcomes(ida);

    node->Info().Header().SetName(name);

    DSL_rectangle &position = node->Info().Screen().position;
    position.center_X = xPos;
    position.center_Y = yPos;
    position.width = 85;
    position.height = 55;

    return handle;
}

File: SMILE/tutorials/java/src/tutorials/Tutorial1.java
package tutorials;

import smile.*;

// Tutorial1 creates a simple network with three nodes,
// then writes its content as XDSL file to disk.

public class Tutorial1 {
    public static void run() {
        System.out.println("Starting Tutorial1...");
        Network net = new Network();
        
        int e = createCptNode(net,
            "Economy", "State of the economy",
            new String[] {"Up","Flat","Down"}, 
            160, 40);

        int s = createCptNode(net,
            "Success", "Success of the venture",
            new String[] {"Success","Failure"}, 
            60, 40);

        int f = createCptNode(net,
            "Forecast", "Expert forecast",
            new String[] {"Good","Moderate","Poor"}, 
            110, 140);

        net.addArc(e, s);
        net.addArc(s, f);
        
        // we can also use node identifiers when creating arcs  
        net.addArc("Economy", "Forecast");
        
        double[] economyDef = {
            0.2, // P(Economy=U)
            0.7, // P(Economy=F)
            0.1  // P(Economy=D)
        };
        net.setNodeDefinition(e, economyDef);
        
        double[] successDef = new double[] {
            0.3, // P(Success=S|Economy=U)
            0.7, // P(Success=F|Economy=U)
            0.2, // P(Success=S|Economy=F)
            0.8, // P(Success=F|Economy=F)
            0.1, // P(Success=S|Economy=D)
            0.9  // P(Success=F|Economy=D)
        };
        net.setNodeDefinition(s, successDef);
        
        
        double[] forecastDef = new double[] {
            0.70, // P(Forecast=G|Success=S,Economy=U)
            0.29, // P(Forecast=M|Success=S,Economy=U)
            0.01, // P(Forecast=P|Success=S,Economy=U)
    
            0.65, // P(Forecast=G|Success=S,Economy=F)
            0.30, // P(Forecast=M|Success=S,Economy=F)
            0.05, // P(Forecast=P|Success=S,Economy=F)
    
            0.60, // P(Forecast=G|Success=S,Economy=D)
            0.30, // P(Forecast=M|Success=S,Economy=D)
            0.10, // P(Forecast=P|Success=S,Economy=D)
    
            0.15, // P(Forecast=G|Success=F,Economy=U)
            0.30, // P(Forecast=M|Success=F,Economy=U)
            0.55, // P(Forecast=P|Success=F,Economy=U)
    
            0.10, // P(Forecast=G|Success=F,Economy=F)
            0.30, // P(Forecast=M|Success=F,Economy=F)
            0.60, // P(Forecast=P|Success=F,Economy=F)
    
            0.05, // P(Forecast=G|Success=F,Economy=D)
            0.25, // P(Forecast=G|Success=F,Economy=D)
            0.70  // P(Forecast=G|Success=F,Economy=D)	
        };
        net.setNodeDefinition(f, forecastDef);

        net.writeFile("tutorial1.xdsl");

        System.out.println(
            "Tutorial1 complete: Network written to tutorial1.xdsl");
    }
    
    
    private static int createCptNode(
            Network net, String id, String name, 
            String[] outcomes, int xPos, int yPos) {
        int handle = net.addNode(Network.NodeType.CPT, id);
        
        net.setNodeName(handle, name);
        net.setNodePosition(handle, xPos, yPos, 85, 55);
        
        int initialOutcomeCount = net.getOutcomeCount(handle); 
        for (int i = 0; i < initialOutcomeCount; i ++) {
            net.setOutcomeId(handle, i, outcomes[i]);
        }
        
        for (int i = initialOutcomeCount; i < outcomes.length; i ++) {
            net.addOutcome(handle, outcomes[i]);
        }
        
        return handle;
    }
}
